src/main/java/org/broadinstitute/consent/http/service/DatasetService.java

```diff
@@ -57,6 +61,46 @@ public DataSet getDatasetWithPropertiesById(Integer datasetId) {
         return dataset;
     }
 
+    public Optional<DataSet> updateDataset(DataSetDTO dataset, Integer datasetId, Integer userId) {
+        Timestamp now = new Timestamp(new Date().getTime());
+
+        DataSet old = getDatasetWithPropertiesById(datasetId);
+        Set<DataSetProperty> oldProperties = old.getProperties();
+
+        List<DataSetPropertyDTO> updateDatasetPropertyDTOs = dataset.getProperties();
+        List<DataSetProperty> updateDatasetProperties = processDatasetProperties(datasetId, updateDatasetPropertyDTOs);
+
+        List<DataSetProperty> propertiesToAdd = updateDatasetProperties.stream()
+              .filter(p -> oldProperties.stream()
+                    .noneMatch(op -> op.getPropertyKey() == p.getPropertyKey()))
+              .collect(Collectors.toList());
+
+        List<DataSetProperty> propertiesToUpdate = updateDatasetProperties.stream()
+              .filter(p -> oldProperties.stream()
+                    .noneMatch(op -> p.equals(op)))
+              .collect(Collectors.toList());
+
+        List<DataSetProperty> propertiesToDelete = oldProperties.stream()
+              .filter(op -> updateDatasetProperties.stream()
+                .noneMatch(p -> p.getPropertyKey() == op.getPropertyKey())
+              ).collect(Collectors.toList());
+
+        if (propertiesToAdd.isEmpty() && propertiesToUpdate.isEmpty() && propertiesToDelete.isEmpty()) {
+            return Optional.empty();
+        }
+
+        updateDatasetProperties(datasetId, propertiesToUpdate, propertiesToDelete, propertiesToAdd);
+        dataSetDAO.updateDatasetUpdateUserAndDate(datasetId, now, userId);
+        DataSet updatedDataset = getDatasetWithPropertiesById(datasetId);
+        return Optional.of(updatedDataset);
+    }
+
+    private void updateDatasetProperties(Integer datasetId, List<DataSetProperty> updateProperties, List<DataSetProperty> deleteProperties, List<DataSetProperty> addProperties) {
```

Comment:
nit: minor possible optimization here ... `DataSetProperty` has a dataset id field. If that were populated before arriving at this method, you wouldn't need to pass in the id at all, just read them from the property.

Reply:
Nice catch!